test/integration/lib/render.js

```diff
@@ -49,22 +52,48 @@ tape.onFinish(() => {
     mapboxgl.clearPrewarmedResources();
 });
 
+let osIgnore;
+let timeout = 5000;
+
+if (process.env.CI) {
+    // On CI, MacOS and Windows run on virtual machines.
+    // These run slower so we increase the timeout.
+    const os = navigator.appVersion;
+    if (os.includes("Macintosh")) {
+        osIgnore = ignoreMac;
+        timeout = 20000;
+    } else if (os.includes("Linux")) {
+        osIgnore = null;
+    } else if (os.includes("Windows")) {
+        osIgnore = ignoreWindows;
+        timeout = 60000;
+    } else { console.warn("Unrecognized OS:", os); }
+}
+
 for (const testName in fixtures) {
-    tape(testName, {timeout: 20000}, ensureTeardown);
+    tape(testName, {timeout}, ensureTeardown);
 }
 
 function ensureTeardown(t) {
     const testName = t.name;
-    const options = {timeout: 5000};
-    if (testName in ignores) {
-        const ignoreType = ignores[testName];
-        if (/^skip/.test(ignoreType)) {
+    const options = {timeout};
+
+    function checkIgnore(ignoreConfig) {
+        if (ignoreConfig.skip.includes(testName)) {
             options.skip = true;
-        } else {
+        }
+
+        if (ignoreConfig.todo.includes(testName)) {
             options.todo = true;
         }
     }
-
+    checkIgnore(ignores);
+    if (osIgnore) {
+        checkIgnore(osIgnore);
+    }
+    if (navigator.userAgent.includes("Firefox")) {
+        checkIgnore(ignoreFirefox);
+    }
```

Comment:
Why is Firefox having a different ignore path instead of assigning `osIgnore` further up?

Reply:
Firefox is a browser, not an OS. (Counterintuitively, `navigator.userAgent` returns a browser string and `navigator.appVersion` returns an OS string.)

My thinking was that we should have independent ignore files for browser and OS that are used together. For instance, if we run a test for Firefox on Windows, it will ignore tests listed in both`firefox.js` and `windows.js`.

Currently this doesn't happen in practice since on CI is all test runs use Chrome except for the firefox linux test.

Reply:
Ah right, I got myself confused between the two. All good then!